### PR TITLE
Add importmap minification

### LINKED ISSUE #169: Support <script type="importmap">
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap

```html
<script type="importmap">
  {
    "imports": {
      "shapes/": "./module/shapes/",
      "othershapes/": "https://example.com/modules/shapes/"
    }
  }
</script>
```
⇓
```html
<script type="importmap">{"imports":{"shapes/":"./module/shapes/","othershapes/":"https://example.com/modules/shapes/"}}</script>
```

Basically `JSON.stringify(JSON.parse(scriptElement.textContent))`.

### PR DESCRIPTION
**Pull request by sdavids**
fixes #169

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,7 +97,7 @@ Most of the options are disabled by default.
 | `preserveLineBreaks`           | Always collapse to 1 line break (never remove it entirely) when whitespace between tags include a line break. Must be used in conjunction with `collapseWhitespace=true` | `false` |
 | `preventAttributesEscaping`    | Prevents the escaping of the values of attributes | `false` |
 | `processConditionalComments`   | Process contents of conditional comments through minifier | `false` |
-| `processScripts`               | Array of strings corresponding to types of script elements to process through minifier (e.g. `text/ng-template`, `text/x-handlebars-template`, etc.) | `[ ]` |
+| `processScripts`               | Array of strings corresponding to types of script elements to process through minifier (e.g. `text/ng-template`, `text/x-handlebars-template`, `importmap`, etc.) | `[ ]` |
```
Comment (DanielRuf) [inline README.md:100]: Personally I think we don't have to mention the type there.

Comment (DanielRuf) [inline README.md:100]: Personally I think we don't have to mention the type there.

File: src/htmlminifier.js [hunk 2]
```diff
@@ -843,6 +845,17 @@ async function createSortFns(value, options, uidIgnore, uidAttr) {
   }
 }
 
+function minifyImportmap(value, options) {
+  if (options.collapseWhitespace) {
+    try {
+      return JSON.stringify(JSON.parse(value));
```
Comment (DanielRuf) [inline src/htmlminifier.js:851]: I'm surprised that we need this approach even if we already can minify JSON-LD / microdata.

See also:
https://github.com/terser/html-minifier-terser/issues/146
https://github.com/terser/html-minifier-terser/issues/34
https://user-images.githubusercontent.com/827205/224982981-01f104bd-4033-4ac5-a22c-706c48c4199c.png

The difference to the available option is not quite clear to me:

grafik

Can you clarify the difference and why we need this extra function here?

Comment (sdavids) [reply src/htmlminifier.js:851]:
```shell
$ printf '
<script type="importmap">
  {
    "imports": {
      "shapes/": "./module/shapes/",
      "othershapes/": "https://example.com/modules/shapes/"
    }
  }
</script>
' | npx html-minifier-terser --process-scripts 'importmap' --collapse-whitespace
```

```html
<script type="importmap">{ "imports": { "shapes/": "./module/shapes/", "othershapes/": "https://example.com/modules/shapes/" } }</script>
```

with the patch you will get:

```html
<script type="importmap">{"imports":{"shapes/":"./module/shapes/","othershapes/":"https://example.com/modules/shapes/"}}</script>
```

Notice the minified JSON.

```shell
$ printf '
<script type="application/ld+json">
{
  "@context": "https://json-ld.org/contexts/person.jsonld",
  "@id": "http://dbpedia.org/resource/John_Lennon",
  "name": "John Lennon",
  "born": "1940-10-09",
  "spouse": "http://dbpedia.org/resource/Cynthia_Lennon"
}
</script>
' | npx html-minifier-terser --process-scripts 'application/ld+json' --collapse-whitespace
```

```html
<script type="application/ld+json">{ "@context": "https://json-ld.org/contexts/person.jsonld", "@id": "http://dbpedia.org/resource/John_Lennon", "name": "John Lennon", "born": "1940-10-09", "spouse": "http://dbpedia.org/resource/Cynthia_Lennon" }</script>
```

I could adjust the patch to also route `application/ld+json` (or everything with `+json` at the end) into `minifyImportmap` -- in that case I would rename it to `minifyJSON` though.

Another alternative would be to introduce a new CLI argument `--process-json-scripts` (or `--process-scripts-json`) which would handle both `importmap` and `application/ld+json` for now.

The implementation would have an array with `importmap` and `application/ld+json` and we would add other micro formats on demand to that array.

Or the CLI parameter could take a argument and behave similar to `--process-scripts` in that you explicitly have to state them.

If we add a new CLI parameter we have to decide what happens when `--process-scripts` also contains the script type, in my opinion it should ignore it.

Comment (DanielRuf) [reply src/htmlminifier.js:851]: I don't think this is the right approach. We should try to find a better way to handle the spaces (which seem to be the only difference).

Comment (sdavids) [reply src/htmlminifier.js:862]: I added some more tests, I will look at a different approach for processScript over the weekend. The documentation has to be updated after ...

Comment (sdavids) [reply src/htmlminifier.js:860]: I have improved the tests and updated the documentation.

`JSON.stringify` in combination with `JSON.parse` does more than just remove spaces:

```shell
$ node -e "console.log(JSON.stringify(JSON.parse('{ \"key\": \"ignored\", \"key\": -2.1E+35}')))"
{"key":-2.1e+35}
```

`{ "key": "ignored", "key": -2.1E+35}`
`{ "key": -2.1e+35}`

⇒ the duplicate key has been removed and the `E` has been lower-cased

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description

Comment (sdavids) [reply src/htmlminifier.js:860]: We could have one option just to remove the whitespaces and another to use `JSON.stringify`.

**Comment by DanielRuf** [comment]
I very much appreciate your effort and the amount of work that you have invested.
Thank you very much.

Still, I think this is not going into the right direction and leads to more edge cases and all this just to get rid of extra spaces.

We just need an **option to clear these extra spaces** (in general), for all inputs (nut just JSON or importmap). The JSON.parse/stringify workaround is not an ideal solution.

Regarding tests, I think real examples make more sense.

**Pull request closed without merging**